test/integration/array_test.go

```diff
@@ -471,3 +471,293 @@ func TestArraySet(t *testing.T) {
 		})
 	}
 }
+
+func TestArrayConcurrencyTable(t *testing.T) {
+	clients := activeClients(t, 2)
+	c0, c1 := clients[0], clients[1]
+	defer deactivateAndCloseClients(t, clients)
+
+	initArr := []int{1, 2, 3, 4}
+	initMarshal := `[1,2,3,4]`
+	oneIdx := 1
+	otherIdxs := []int{2, 3}
+	newValues := []int{5, 6}
+
+	type arrayOp struct {
+		opName   string
+		executor func(*json.Array, int)
+	}
+
+	operations := []arrayOp{
+		// insert
+		{"insert.prev", func(a *json.Array, cid int) {
+			a.InsertIntegerAfter(oneIdx, newValues[cid])
+		}},
+		{"insert.prev.next", func(a *json.Array, cid int) {
+			a.InsertIntegerAfter(oneIdx-1, newValues[cid])
+		}},
+
+		// move
+		{"move.prev", func(a *json.Array, cid int) {
+			a.MoveAfterByIndex(oneIdx, otherIdxs[cid])
+		}},
+		{"move.prev.next", func(a *json.Array, cid int) {
+			a.MoveAfterByIndex(oneIdx-1, otherIdxs[cid])
+		}},
+		{"move.target", func(a *json.Array, cid int) {
+			a.MoveAfterByIndex(otherIdxs[cid], oneIdx)
+		}},
+
+		// set by index
+		{"set.target", func(a *json.Array, cid int) {
+			a.SetInteger(oneIdx, newValues[cid])
+		}},
+
+		// remove
+		{"remove.target", func(a *json.Array, cid int) {
+			a.Delete(oneIdx)
+		}},
+	}
+
+	ctx := context.Background()
+	d0 := document.New(helper.TestDocKey(t))
+	assert.NoError(t, c0.Attach(ctx, d0))
+	d1 := document.New(helper.TestDocKey(t))
+	assert.NoError(t, c1.Attach(ctx, d1))
+
+	runTest := func(op1, op2 arrayOp) testResult {
+		assert.NoError(t, d0.Update(func(root *json.Object, p *presence.Presence) error {
+			root.SetNewArray("a").AddInteger(initArr...)
+			assert.Equal(t, initMarshal, root.GetArray("a").Marshal())
+			return nil
+		}))
+
+		assert.NoError(t, c0.Sync(ctx))
+		assert.NoError(t, c1.Sync(ctx))
+
+		assert.NoError(t, d0.Update(func(root *json.Object, p *presence.Presence) error {
+			op1.executor(root.GetArray("a"), 0)
+			return nil
+		}))
+
+		assert.NoError(t, d1.Update(func(root *json.Object, p *presence.Presence) error {
+			op2.executor(root.GetArray("a"), 1)
+			return nil
+		}))
+
+		flag := syncClientsThenCheckEqual(t, []clientAndDocPair{{c0, d0}, {c1, d1}})
+		if flag {
+			return testResult{flag, `pass`}
+		}
+		return testResult{flag, `different result`}
+	}
+
+	for i1, op1 := range operations {
+		for i2, op2 := range operations {
+			if i1 > i2 {
+				continue
+			}
```

Comment:
I tested only the upper diagonal to reduce test time and avoid symmetric duplicate results. 

From the point of view of convergence or not, I thought that only the order in which the operation is applied was important, and the 'timestamp' of the operation was not.
Although the timestamp might influence the convergence results, I didn't think it would cause only one client to perform different actions. When tested locally, this is probably true in current array operations. 

However, considering the addition of new operations, it seems better to inspect the entire matrix area, even if it slightly increases the test time.

Reply:
Update: I found asymmetric case between Insert and SetByIndex. The tests have to check timestamp as well. 

```
=== RUN   TestArrayConcurrencyTable/insert.prev.next_vs_set.target
2024-09-02T15:43:59.435+0900    INFO    r67     SYNC: 'TestArrayConcurrencyTable' is synced by 'craltbjkssro7f52963g', push: 1, pull: 0, min: 36:4294967295:07  {"prj": "default"}
2024-09-02T15:43:59.445+0900    INFO    r68     SYNC: 'TestArrayConcurrencyTable' is synced by 'craltbjkssro7f529640', push: 0, pull: 1, min: 37:4294967295:0a  {"prj": "default"}
before d1: {"a":[1,5,2,3,4]}
2024-09-02T15:43:59.457+0900    INFO    r69     SYNC: 'TestArrayConcurrencyTable' is synced by 'craltbjkssro7f52963g', push: 1, pull: 0, min: 37:4294967295:0a  {"prj": "default"}
before d2: {"a":[1,6,3,4]}
2024-09-02T15:43:59.465+0900    INFO    b40     SNAP: 'TestArrayConcurrencyTable', serverSeq: 40
2024-09-02T15:43:59.470+0900    INFO    r70     SYNC: 'TestArrayConcurrencyTable' is synced by 'craltbjkssro7f529640', push: 1, pull: 1, min: 39:4294967295:07  {"prj": "default"}
2024-09-02T15:43:59.481+0900    INFO    r71     SYNC: 'TestArrayConcurrencyTable' is synced by 'craltbjkssro7f52963g', push: 0, pull: 1, min: 39:4294967295:07  {"prj": "default"}
after d1: {"a":[1,5,6,3,4]}
after d2: {"a":[1,6,5,3,4]}
    array_test.go:564: different result

// ...

=== RUN   TestArrayConcurrencyTable/set.target_vs_insert.prev.next
2024-09-02T12:16:10.825+0900    INFO    r187    SYNC: 'TestArrayConcurrencyTable' is synced by 'crairu3kssrjf9q0saug', push: 1, pull: 0, min: 108:4294967295:f5 {"prj": "default"}
2024-09-02T12:16:10.836+0900    INFO    r188    SYNC: 'TestArrayConcurrencyTable' is synced by 'crairu3kssrjf9q0sav0', push: 0, pull: 1, min: 109:4294967295:f8 {"prj": "default"}
before d1: {"a":[1,5,3,4]}
2024-09-02T12:16:10.849+0900    INFO    r189    SYNC: 'TestArrayConcurrencyTable' is synced by 'crairu3kssrjf9q0saug', push: 1, pull: 0, min: 109:4294967295:f8 {"prj": "default"}
before d2: {"a":[1,6,2,3,4]}
2024-09-02T12:16:10.862+0900    INFO    r190    SYNC: 'TestArrayConcurrencyTable' is synced by 'crairu3kssrjf9q0sav0', push: 1, pull: 1, min: 111:4294967295:f5 {"prj": "default"}
2024-09-02T12:16:10.875+0900    INFO    r191    SYNC: 'TestArrayConcurrencyTable' is synced by 'crairu3kssrjf9q0saug', push: 0, pull: 1, min: 111:4294967295:f5 {"prj": "default"}
after d1: {"a":[1,6,5,3,4]}
after d2: {"a":[1,6,5,3,4]}
```

Reply:
Yes, when I ran the test, it was as follows. 
![image](https://github.com/user-attachments/assets/b3be3e03-9fa9-47dc-9489-82a2ec465386)

Reply:
It seems to that asymmetric divergence only occurs between small-timestamp positional operation and large-timestamp non-positional operation such that `node.next.PositionedAt().After(executedAt)` satisfies.

Reply:
Without Move:
<img width="724" alt="Screenshot 2024-09-02 at 5 54 45 PM" src="https://github.com/user-attachments/assets/47c34b0a-a854-4a0e-ac08-0e751658e967">